NE-2186: Propagate ingress labels to routes

rikatz opened the pull request:
This change adds the propagation of ingress resource labels to route resources.

This behavior is controller by the annotation route.openshift.io/reconcile-labels and once the annotation is enabled, the sync process will patch the child route resource and replace the labels on it for the same labels of the parent ingress.

In case the annotation is removed later, the old labels will be kept but no further replacement will happen.

This fix is based on https://github.com/openshift/route-controller-manager/commit/b48eb95de9652b8102d667f415a9e22c59c6aa6c but adding a flag to not enable the sync directly

rikatz committed:
[NE-2186] Propagate ingress labels to routes

This change adds the propagation of ingress resource labels
to route resources.

This behavior is controller by the annotation route.openshift.io/reconcile-labels
and once the annotation is enabled, the sync process will patch
the child route resource and replace the labels on it for the
same labels of the parent ingress.

In case the annotation is removed later, the old labels will
be kept but no further replacement will happen

rikatz commented:
Per the lack of e2e so far, I have executed the tests manually:
--- During updates ---
* Create an ingress without the annotation
* Add a label to ingress - not replicated to route
* Add the flag annotation to the ingress - labels are replicated to route
* Add a new label to ingress - replicated to route
* Remove annotation from ingress 
* Add a new label to ingress - not replicated
* Add a new label to route - not replaced
* Add the flag annotation back to ingress - labels on route are replaced by the same labels of ingress

--- During a new object creation without the annotation ---
* Create a new object already with labels - route gets the replicated labels (this is the current behavior already)
* Add a new label to ingress - not replicated to route (no annotation)

--- During a new object creation with the annotation ---
* Create a new object already with labels and the annotation flag - route gets the replicated labels (this is the current behavior already)
* Add a new label to ingress - not replicated to route (no annotation)

rikatz commented:
/retest-required

rikatz commented:
same failure, different repo: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1294/pull-ci-openshift-cluster-ingress-operator-master-e2e-hypershift/1978807569782673408

seems something is broken on hypershift test

Miciah commented on pkg/route/ingress/ingress.go:
```diff
@@ -508,8 +508,19 @@ func (c *Controller) sync(key queueKey) error {
 		}
+
+		labelReplacementString := ""
+		if propagateLabelsToRoute(route.GetAnnotations()) {
```
We use `route.Annotations` elsewhere:
https://github.com/openshift/route-controller-manager/blob/7d91acb4f78f24a0cc12479bedc7fbf146df1ae6/pkg/route/ingress/ingress.go#L503
https://github.com/openshift/route-controller-manager/blob/7d91acb4f78f24a0cc12479bedc7fbf146df1ae6/pkg/route/ingress/ingress.go#L735
Any reason not to be consistent here?

rikatz replied:
old habits, will change it here

Miciah commented on pkg/route/ingress/ingress.go:
```diff
@@ -508,8 +508,19 @@ func (c *Controller) sync(key queueKey) error {
 			`{"op":"replace","path":"/metadata/annotations","value":%s},`+
+			labelReplacementString+
 			`{"op":"replace","path":"/metadata/ownerReferences","value":%s}]`,
 			data, annotations, ownerRefs))
```
Adding `labelReplacementString+` presumably changes this string concatenation from a compile-time operation (concatenating constant strings) into a run-time operation (concatenating constants and variables).  What do you think about using string interpolation instead?
```suggestion
			`%s`+
			`{"op":"replace","path":"/metadata/ownerReferences","value":%s}]`,
			data, annotations, labelReplacementString, ownerRefs))
```

More radically, this whole `data = []byte(fmt.Sprintf(` statement is getting complicated.  What do you think about making a list of patches (`var patches []string`, `patches = append(patches, annotationsPatch)`, etc.) and joining them at the end (`data = []byte(fmt.Sprintf("[%s]", strings.Join(patches, ",")))`?

Miciah commented on pkg/route/ingress/ingress.go:
```diff
@@ -714,14 +725,18 @@ func routeMatchesIngress(
 		reflect.DeepEqual(route.Annotations, ingress.Annotations) &&
-		route.OwnerReferences[0].APIVersion == "networking.k8s.io/v1"
+		route.OwnerReferences[0].APIVersion == "networking.k8s.io/v1" &&
+		// Labels are flagged. If the propagation to labels is disabled, we don't care about them matching,
```
I'm not sure what you mean by "flagged".  Do you mean something like "conditionalized", as in "Matching labels is conditional on the 'reconcile-labels' annotation's being set to 'true'"?

rikatz replied:
flagged - like a feature flag :) , gated.

I can change the comment here to "label replication is conditional to the enablement of propagation" wdyt?

Miciah commented on pkg/route/ingress/ingress.go:
```diff
@@ -973,3 +988,8 @@ func destinationCACertificateForIngress(ingress *networkingv1.Ingress, secretLis
+
+func propagateLabelsToRoute(annotations map[string]string) bool {
+	propagateLabelsFlag, ok := annotations[routecontroller.PropagateIngressLabelFlag]
+	return ok && propagateLabelsFlag == "true"
```
This seems likely to cause confusion for values such as "True", "on", or the empty string.  Would it make sense to recognize more values as indicating that labels should be propagated, or emit an ~error~ event for any value other than "true" or "false"?

rikatz replied:
tbh after implementing i didn't liked much the usage of "true" here. A user creating an annotation editing the manifest with something like "annotation: true" will get an error because yaml parser will consider it a boolean and not a string.

eg.:
```
# The edited file had a syntax error: unable to decode "edited-file": json: cannot unmarshal bool into Go struct field ObjectMeta.metadata.annotations of type string
....
metadata:
  annotations:
    xxx: "true"
    blabla: true
```

Maybe we should use something else instead of true/false here, we can discuss about it

Miciah commented on pkg/routecontroller/wellknown.go:
```diff
@@ -14,6 +14,8 @@ var (
 	// TerminationPolicyAnnotationKey defines what TLSTerminationType should be used
 	// on the Route being created
 	TerminationPolicyAnnotationKey = routev1.GroupName + "/termination"
+	//
```
Missing godoc.

Miciah commented:
/retitle [NE-2186](https://issues.redhat.com//browse/NE-2186): Propagate ingress labels to routes

Miciah commented on pkg/route/ingress/ingress_test.go:
```diff
@@ -1326,6 +1406,150 @@ func TestController_sync(t *testing.T) {
+								"route.openshift.io/reconcile-labels": "true",
+							},
+							Labels: map[string]string{
+								"my-label": "somevalue",
+							},
+						},
+						Spec: networkingv1.IngressSpec{
+							Rules: []networkingv1.IngressRule{
+								{
+									Host: "test.com",
+									IngressRuleValue: networkingv1.IngressRuleValue{
+										HTTP: &networkingv1.HTTPIngressRuleValue{
+											Paths: []networkingv1.HTTPIngressPath{
+												{
+													Path:     "/",
+													PathType: &pathTypePrefix,
+													Backend: networkingv1.IngressBackend{
+														Service: &networkingv1.IngressServiceBackend{
+															Name: "service-1",
+															Port: networkingv1.ServiceBackendPort{
+																Name: "http",
+															},
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				}},
+				r: &routeLister{Items: []*routev1.Route{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:            "1-abcdef",
+							Namespace:       "test",
+							OwnerReferences: []metav1.OwnerReference{{APIVersion: "networking.k8s.io/v1", Kind: "Ingress", Name: "1", Controller: &boolTrue}},
+						},
```
First, I'll note, this test case could represent the case where the label was added to the ingress object after the route had been generated, or the case where the label was deleted from the generated route object.  The controller watches ingresses and routes, so in either case, the controller should synchronize the labels by patching the route with the missing label.  

Next, I'll suggest, it would make sense to add another test case to represent the cases where the label is deleted from the ingress object (in which case it should be deleted from the route) or added to the route object (in which case also it should be deleted from the route).

rikatz replied:
added a new test

rikatz committed:
[NE-2186] Structure patch to be applied

This commit changes the way we generate the patch to be
applied to a route, from a unstructured array of bytes
to a better structured map[string]interface{} containing
the real structures, and calling the json marshaller
a single time

rikatz commented on pkg/route/ingress/ingress.go:
```diff
@@ -496,22 +496,36 @@ func (c *Controller) sync(key queueKey) error {
-		if err != nil {
-			return err
+
+		if propagateLabelsToRoute(route.Annotations) {
```
@Miciah wdyt about this new approach? 

Also, take a look into unit tests, the only changed part is the assertion for the label patch, as we are adding it in the end

rikatz committed:
[NE-2186] Add a test case on deleting labels from route

rikatz committed:
[NE-2186] Emit event on invalid annotation value

This change emits an event in case the reconcile-labels contains
an invalid value.

Additionally it adds this assertion to the unit test, checking if
the event was correctly issued to the event recorder.

Miciah commented on pkg/route/ingress/ingress.go:
```diff
@@ -415,6 +416,15 @@ func (c *Controller) sync(key queueKey) error {
 
+	propagateLabels, err := shouldPropagateLabelsToRoute(ingress.Annotations)
+	if err != nil {
+		c.eventRecorder.Eventf(&corev1.ObjectReference{
+			Kind:      "Ingress",
+			Namespace: key.namespace,
+			Name:      key.name,
+		}, corev1.EventTypeNormal, "InvalidAnnotationValue", "Invalid value on annotation %q", routecontroller.PropagateIngressLabelFlag)
+	}
```
Does it make sense to continue in the case of an error?  

I guess there is an implicit assumption that `propagateLabels` will be `false` in this case, and then continuing means we keep reconciling everything other than the labels; that seems like a reasonable design decision.  It would be good to add a comment to make your intention explicit.

rikatz replied:
yeah, IMO it does make sense to continue. I will add a comment

Miciah commented on pkg/route/ingress/ingress.go:
```diff
@@ -991,7 +1002,11 @@ func destinationCACertificateForIngress(ingress *networkingv1.Ingress, secretLis
+func shouldPropagateLabelsToRoute(annotations map[string]string) (bool, error) {
 	propagateLabelsFlag, ok := annotations[routecontroller.PropagateIngressLabelFlag]
-	return ok && propagateLabelsFlag == "true"
+	if !ok || propagateLabelsFlag == "" {
```
I expected the empty value would be treated as an error, rather than assuming it means "false".

rikatz replied:
fixed

Miciah commented on pkg/route/ingress/ingress.go:
```diff
@@ -991,7 +1002,11 @@ func destinationCACertificateForIngress(ingress *networkingv1.Ingress, secretLis
+		return false, nil
+	}
+
+	return strconv.ParseBool(strings.ToLower(propagateLabelsFlag))
```
From a quick glance at [`strconv.ParseBool`'s documentation](https://pkg.go.dev/strconv#ParseBool), its first return value is undefined in case of an error.  However, your changes to `sync` seems to be written with the assumption that `shouldPropagateLabelsToRoute` returns `false` in case of an error.  So I have two suggestions:

- Explicitly check `strconv.ParseBool`'s return value and return `false, err` if `err != nil`.
- Add godoc to `shouldPropagateLabelsToRoute` with an explicit contract that it returns `false` if `err != nil`.

rikatz replied:
Actually because ParseBool doesn't return a pointer, but a strict value, it MUST return false/true, which in fact happens under the hood: 

https://cs.opensource.google/go/go/+/refs/tags/go1.25.3:src/strconv/atob.go;l=17

Anyway, it does make sense to explicitly return "false" so in case the behavior of ParseBool changes in the future (I dunno, but...) we have the explicit expected behavior on our function

Miciah commented on pkg/route/ingress/ingress_test.go:
```diff
@@ -4294,6 +4299,19 @@ func TestController_sync(t *testing.T) {
+				events = append(events, i)
+			}
+
+			for _, expectedEvent := range tt.expectedEvents {
+				if !slices.Contains(events, expectedEvent) {
+					t.Errorf("event %q not found on recorded events", expectedEvent)
+				}
+			}
```
Should we also verify that no *un*expected events were recorded?  Of course, then you would need to add `expectedEvents` for some existing test cases, which is a bit of scope creep, so I don't object if you prefer not to do that in this PR.

rikatz replied:
it is fine, I can fix it quickly here. I will make an assertion if two slices (recorded events and expected events) are exactly the same, this way we assert if all the events we expect were registered, and in the right order

rikatz replied:
fixed

Miciah commented on pkg/route/ingress/ingress.go:
```diff
@@ -496,7 +506,7 @@ func (c *Controller) sync(key queueKey) error {
 	// update any existing routes in place
 	for _, route := range updates {
-		patchOperations := []map[string]interface{}{
+		patchOperations := []map[string]any{
```
I expect this change was meant to be part of https://github.com/openshift/route-controller-manager/commit/225fb3b3ce0768cbf95a8f3413533d21cc65ff0b.

rikatz replied:
yeah, I let this slip into the latest commit, will rebase here and make this part of the implementation

rikatz replied:
moved to the right commit

rikatz committed:
[NE-2186] Fix reviews on shouldPropagateLabelsToRoute

This commit improves the godoc of shouldPropagateLabelsToRoute function, adding
the right contract to it and explicitly returning false in case there is an error
when parsing the boolean.
Additionally adds a comment on the call of this function, to make it clear that
an error when it is being called does not represent a reconciliation failure,
so an event will be registered but the reconciliation will continue

rikatz committed:
[NE-2186] Add event recording test

This commit adds proper event recording test to ingress sync.
Because we now verify all of the events that are registered, and if
the registered is EXACTLY the same as expected, the definition of
other tests was also updated to reflect the expected events

candita commented:
/assign @Miciah @davidesalerno

rikatz commented:
/test okd-scos-e2e-aws-ovn

davidesalerno commented on pkg/route/ingress/ingress.go:
```diff
@@ -973,3 +1004,22 @@ func destinationCACertificateForIngress(ingress *networkingv1.Ingress, secretLis
+		return false, nil
+	}
+
+	shouldPropagate, err := strconv.ParseBool(strings.ToLower(propagateLabelsFlag))
```
```suggestion
	shouldPropagate, err := strconv.ParseBool(strings.TrimSpace(propagateLabelsFlag))
```
`strconv.ParseBool ` supports mixed case, if we would like to be safe we can trim whitespaces to avoid errors in case of ' true '

rikatz replied:
good call, let me do it

davidesalerno commented on pkg/route/ingress/ingress.go:
```diff
@@ -415,6 +416,18 @@ func (c *Controller) sync(key queueKey) error {
+			Kind:      "Ingress",
+			Namespace: key.namespace,
+			Name:      key.name,
+		}, corev1.EventTypeNormal, "InvalidAnnotationValue", "Invalid value on annotation %q", routecontroller.PropagateIngressLabelFlag)
```
```suggestion
		}, corev1.EventTypeNormal, "InvalidAnnotationValue", "Invalid value on annotation %q due to: %q", routecontroller.PropagateIngressLabelFlag, err)
```

It could be useful to add the error in the event to help diagnostic and debugging

rikatz committed:
[NE-2186] Add better error message and bool parsing

This commit adds a better error message on an event in case of
failing to parse the new annotation that enables label replication.

It also adds better parsing on the bool value, trimming spaces

davidesalerno commented:
/lgtm
/approve

rikatz commented:
/cc @rhamini3 @lihongan @ShudiLi @melvinjoseph86 

for QE verification! Thanks!

ShudiLi commented:
Even set "route.openshift.io/reconcile-labels" annotation with an "invalid" value, all  the ingress's annotations are synced to the route, not sure it is expected.
```
1.
% oc get clusterversion
NAME      VERSION                                                AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.21.0-0-2025-11-06-052231-test-ci-ln-dc9w45t-latest   True        False         154m    Cluster version is 4.21.0-0-2025-11-06-052231-test-ci-ln-dc9w45t-latest

2.
% oc -n e2e-test-router-ingressclass-rtm46 get ingress ingress-edge  -oyaml | yq ".metadata.annotations"
cccc: dddd2
route.openshift.io/reconcile-labels: "true"
route.openshift.io/termination: edge

3.
% oc -n e2e-test-router-ingressclass-rtm46 get route ingress-edge-t48qf -oyaml | yq ".metadata.annotations"      
cccc: dddd2
route.openshift.io/reconcile-labels: "true"
route.openshift.io/termination: edge


4.  set the "route.openshift.io/reconcile-labels" annotation with non "true" value 
% oc -n e2e-test-router-ingressclass-rtm46 get ingress ingress-edge -oyaml | yq ".metadata.annotations"
cccc: dddd66
route.openshift.io/reconcile-labels: ffalse
route.openshift.io/termination: edge

5.
% oc -n e2e-test-router-ingressclass-rtm46 get route ingress-edge-t48qf -oyaml | yq ".metadata.annotations"
cccc: dddd66
route.openshift.io/reconcile-labels: ffalse
route.openshift.io/termination: edge
```

rikatz commented:
Hi @ShudiLi yes, it is expected.

In fact originally just the annotations where synced, any annotation you add to ingress is synced to route. 

This change is just about allowing a user to enable the sync of labels from ingress to route, and as this is controlled by an annotation, it is expected that this same annotation is also sent to the route :)

ShudiLi commented:
/label qe-approved
thanks

rikatz commented:
@ShudiLi I guess for this to be mergeable it also needs the "/verified" label, if you don't mind! :) 

Thanks!

rikatz commented:
(I mean I can add the verified by pointing to your comment as well!)

rikatz commented:
/jira refresh

rikatz commented:
Spoken with @alebedev87 here, as it is all approved by QE and engineers I will add the verified by

Thanks all
/verified by @ShudiLi https://github.com/openshift/route-controller-manager/pull/67#issuecomment-3497203782
